[ENH] Implemented `to_dagitty` and `from_dagitty` methods for MAG

JATAYU000 opened the pull request:
### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [ ] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Fixes #2370

### List of changes to the codebase in this pull request
- Methods `to_dagitty` and `from_dagitty` for `MAG` having similar method structure as implemented in `DAG`
- Added Examples for each method
- Update the parser implementation.

JATAYU000 committed:
Implemented to_dagitty and from_dagitty for MAG

JATAYU000 committed:
Removed print and fixed Example code

JATAYU000 committed:
Merge branch 'dev' into feature/MAG_dagitty

ankurankan reviewed (changes requested):
Would it be possible to move both of these methods to the `AncestralBase` class? As both `MAG` and `PAG` inherit `AncestralBase`, if we can generalize the implementation, we should be able to reduce duplicated code. That will also fix both #2370 and #2371.

The `from_dagitty` method will need to be generalized. It makes certain assumptions at the moment, for example, each edge is on a single line. Instead of writing the method from scratch, I would suggest modifying our dagitty parser (`pgmpy.utils.parse_dagitty`) instead to handle `MAG` and `PAG` as well.

JATAYU000 commented:
Understood, Will work on it.

JATAYU000 commented:
@ankurankan is `PAG` already implemented in base? Can you point me where it is implemented. I can only see `MAG` which is being inherited by `AncestralBase`.

ankurankan commented:
@JATAYU000 Sorry, I forgot to mention earlier that it is still in the works: https://github.com/pgmpy/pgmpy/pull/2346/files#diff-f58eaac198e83d52f2277f52b857f97decc963b9bad79300db7bd3f5cccf9aa2R9. The main difference between `MAG` and `PAG` would be that `PAG` would allow for a new edge type where one of the ends of the edge could be unknown, represented using edge type `o`.

JATAYU000 commented:
Makes sense, Thank you

JATAYU000 committed:
Migrated dagitty methods to AncestralBase

JATAYU000 commented:
@ankurankan I have migrated the methods to `AncestralBase`
Now for the `parse_dagitty` utility I am planning on adding a new argument `target_type` which can be 'DAG', 'MAG' etc.

JATAYU000 committed:
Merge branch 'dev' into feature/MAG_dagitty

ankurankan commented:
@JATAYU000 I have fixed the failing test in the `dev` branch. If you update your branch, it should get fixed.

JATAYU000 committed:
Merge branch 'dev' into feature/MAG_dagitty

JATAYU000 commented:
@ankurankan I will start Modifying the `parse_dagitty` utility once #2395 is merged.

JATAYU000 committed:
Merge branch 'dev' into feature/MAG_dagitty

JATAYU000 committed:
Generalised parse_dagitty function

JATAYU000 commented:
@ankurankan Current status of this PR,
- Added `target_type` argument to `parse_dagitty` function
- Instead of returning `latent` it now returns a dict `roles`
- Updated the `DAG.from_dagitty` method to use the new `roles`
- Similarly in `Ancestralbase` class as well.

in `DAG` when there are `coefs` it uses `LinearGaussianBayesianNetwork` and it does not take roles, it has latent alone, I assume it only needs latent. I want your input regarding that part

ankurankan commented on pgmpy/base/AncestralBase.py:
```diff
@@ -528,6 +529,124 @@ def get_reachable_nodes(self, node, u_type=None, v_type=None):
+    def to_dagitty(self) -> str:
+        """
+        Convert the MAG to a Dagitty string representation.
+        Returns
```
There needs to be an empty line before this; otherwise, `sphinx` won't be able to parse the section.

ankurankan commented on pgmpy/base/AncestralBase.py:
```diff
@@ -528,6 +529,124 @@ def get_reachable_nodes(self, node, u_type=None, v_type=None):
+        -----
+        - Node names are converted to string representations.
+        - If node names contain spaces or special characters, they will be used as-is.
+        - Users should ensure node names are valid in R/dagitty context if needed.
```
I would suggest keeping the first statement in the main method description in line 534. I think the rest of the points are a bit obvious, and we can remove the `Notes` section.

ankurankan commented on pgmpy/base/AncestralBase.py:
```diff
@@ -528,6 +529,124 @@ def get_reachable_nodes(self, node, u_type=None, v_type=None):
+        ----------
+        dagitty syntax: https://cran.r-project.org/web/packages/dagitty/dagitty.pdf
+        """
+        dagitty_str = "dag {\n"
```
I think the `dag` in the string will need to be dependent on the class from which the `to_dagitty` method is being called. This can be `mag` or `pag`.

ankurankan commented on pgmpy/base/AncestralBase.py:
```diff
@@ -528,6 +529,124 @@ def get_reachable_nodes(self, node, u_type=None, v_type=None):
+            elif marks[u] == ">" and marks[v] == ">":
+                dagitty_str += f"{u} <-> {v}\n"
+            elif marks[u] == "-" and marks[v] == "-":
+                dagitty_str += f"{u} -- {v}\n"
```
I think this currently doesn't consider the edge marker `o`. dagitty represents the `o` marker as a `@`.

JATAYU000 replied:
Okay so daggity represents it as '@' and we should map it to 'o'

ankurankan replied:
@JATAYU000 Since this is the `AncestralBase` class, I think this should handle the `o` edge marker too. Yes, dagitty represents `o` as `@`

ankurankan commented on pgmpy/utils/parser.py:
```diff
@@ -67,30 +67,33 @@ def parse_lavaan(lines):
 
 
-def parse_dagitty(lines):
+def parse_dagitty(lines, target_type="DAG"):
```
I don't think we would need a `target_type` here. The dagitty syntax already defines the type of model. The dagitty string starts with either `dag`, `mag`, or `pag`. We can use that to determine the class that we want to return.

JATAYU000 commented:
@ankurankan , I have started resolving the changes requested, Ideally we want the parser to be able to parse all the examples in pages 10 and 11 [dagitty pdf](https://cran.r-project.org/web/packages/dagitty/dagitty.pdf) right?
and apart from the requested changes these points below are accepted?

>  Instead of returning latent it now returns a dict roles
>  Updated the DAG.from_dagitty method to use the new roles
>  Similarly in Ancestralbase class as well.

ankurankan committed:
Merge branch 'dev' into feature/MAG_dagitty

ankurankan commented:
@JATAYU000 Sorry for the delay in the reply. I think we will need to write a few of our own examples as the examples shown in the pdf do not cover all the scenarios. 

Also, the tests seem to be failing. Could you please have a look?

JATAYU000 commented:
> I think we will need to write a few of our own examples as the examples shown in the pdf do not cover all the scenarios.

@ankurankan right now the parser works for all the examples except for group operation, It would be helpful if you could point me to some documentation on use of the other scenarios which are not mentioned in the pdf.
Also figured out why the test suites failed, working on it.

ankurankan commented:
@JATAYU000 

>  It would be helpful if you could point me to some documentation on use of the other scenarios which are not mentioned in the pdf.

I don't know of any resources that have a database of MAGs. What I would suggest is that we write test cases that at least checks that all the different edge types are parsed correctly. One edge case to consider is when a node is disconnected from the graph. You can have a look at the DAG parser tests for some examples: https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/tests/test_base/test_DAG.py#L789 

Let me know if you need any help in creating example models to test using, and I can try to write some.

JATAYU000 committed:
Parse pag and mag, Move methods to AncestralBase

JATAYU000 committed:
Merge branch 'feature/MAG_dagitty' of github.com:JATAYU000/pgmpy into feature/MAG_dagitty

JATAYU000 committed:
Merge branch 'dev' into feature/MAG_dagitty

JATAYU000 committed:
Added test for MAG.from_dagitty, updated parser

JATAYU000 commented:
Update on the PR:
- Parser is updated for most of the edge cases which I am aware of including the group operations and disconnected graphs
- Added Tests for MAG with some examples which cover most of the edge cases
- I Did not notice before that  `with_roles(`) only supports `exposure` and `outcome` and not `latents` even though there exists a `test test_latents_with_role` which checks if `with_roles()` adds `latent` role when it does not. This issue exists only in DAG which fails to add `latents` through `roles` dict, unlike MAG which passes the `latents` as params to `super()` . I have decided to fix it by appending the latents provided through the roles as well which is what the `test_latents_with_role` expects.

JATAYU000 committed:
Merge branch 'dev' into feature/MAG_dagitty

ankurankan committed:
Merge branch 'dev' into feature/MAG_dagitty

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -182,6 +182,8 @@ def __init__(
         elif not isinstance(roles, dict):
             raise TypeError("Roles must be provided as a dictionary.")
+        else:
+            self.latents = self.latents.union(roles.get("latents", set()))
```
Why is this required? Isn't the latents role being set in Line 177?

JATAYU000 replied:
No, I was wrong about how `with_roles()` handled latents at that time , It is clear now.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -287,9 +289,13 @@ def from_dagitty(cls, string=None, filename=None) -> "DAG":
 
-        ebunch, latents, coefs, nodes = parse_dagitty(dagitty_str)
+        ebunch, roles, coefs, nodes = parse_dagitty(dagitty_str)
+        latents = roles["latents"]
```
I don't think we will need to extract `latents`. If it is present in the `roles`, it should get set automatically.

JATAYU000 replied:
@ankurankan That was for `LinearGaussianBayesianNetwork` ,sure I can move that inside its block.

JATAYU000 committed:
Removed unnecessary latents set

JATAYU000 committed:
Merge

JATAYU000 commented:
@ankurankan Sorry for the delay but I have resolved the issues, I had some changes which did not set latents variable, I have removed it since it is not required. Is there anything else to be done in this pr? PAG tests with the parser will be added in [This PR](https://github.com/pgmpy/pgmpy/pull/2346) right?

JATAYU000 committed:
Merge branch 'dev' into feature/MAG_dagitty

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -287,16 +287,21 @@ def from_dagitty(cls, string=None, filename=None) -> "DAG":
+            if cls.__name__ == "DAG":
+                dag = cls(ebunch=ebunch, roles=roles)
+            else:
+                dag = cls(ebunch=ebunch, latents=roles["latents"])
```
Why do we need the if-else here? Shouldn't we always pass all the roles?

JATAYU000 replied:
This was left behind before MAG implementation , Forgot to remove it thank you for pointing it out.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -287,16 +287,21 @@ def from_dagitty(cls, string=None, filename=None) -> "DAG":
-            lgbn = LinearGaussianBayesianNetwork(ebunch=ebunch, latents=latents)
+            lgbn = LinearGaussianBayesianNetwork(
+                ebunch=ebunch, latents=roles["latents"]
+            )
```
Shouldn't we pass the roles here as well?

ankurankan replied:
I have changed the API a bit, and now `LinearGaussianBayesianNetwork` accepts the same arguments that `DAG` does. Ref #2455

ankurankan commented on pgmpy/tests/test_base/test_MAG.py:
```diff
@@ -161,3 +163,43 @@ def test_graph_properties(self, mag):
+
+        # Handles order of nodes
+        def normalize_edge(edge):
+            return tuple(sorted(edge))
```
I think this is incorrect. For a directed edge, it can potentially flip the direction.

ankurankan commented on pgmpy/tests/test_base/test_MAG.py:
```diff
@@ -161,3 +163,43 @@ def test_graph_properties(self, mag):
+        model_from_file = MAG.from_dagitty(filename="test_model.dagitty")
+        os.remove("test_model.dagitty")
+
+        expected_edges = {("A", "B,"), ("A", "E"), ("A", "J"), ("A", "M")}
```
I think there is a mistake in the expected edges here. In the first edge `("A", "B,")`, the comma is inside the quotes.

ankurankan reviewed (changes requested):
@MAVRICK-1 Sorry for the delay in the review, it somehow fell off my list of PRs to review.

JATAYU000 commented:
> I think there is a mistake in the expected edges here. In the first edge ("A", "B,"), the comma is inside the quotes.

That was a mistake in the group item parsing which was responsible for the normalize edge in the tests, I have fixed it and updated the tests

> Shouldn't we pass the roles here as well?
```py
class LinearGaussianBayesianNetwork(DAG):
    def __init__(
        self,
        ebunch: Optional[List[Tuple[Hashable, Hashable]]] = None,
        latents: Set[Hashable] = set(),
        lavaan_str: Optional[str] = None,
        dagitty_str: Optional[str] = None,
    ) -> None:
        super(LinearGaussianBayesianNetwork, self).__init__(
            ebunch=ebunch,
            latents=latents,
        )
```
@ankurankan  Why does it not call super with `roles` and just the `latents` then? Is it because it does not benefit from other roles? Should I update here to pass `roles` as well?

Also idk why these doctest errors are there now , The doctest are failing in dev as well. Are you aware of this Issue?

JATAYU000 committed:
Fixed group operation bug and tests for it

JATAYU000 committed:
Merge

JATAYU000 committed:
Merge branch 'dev' into feature/MAG_dagitty

ankurankan committed:
Merge branch 'dev' into feature/MAG_dagitty

ankurankan reviewed (changes requested):
@JATAYU000 I have added a couple of minor comments; looks good otherwise.

JATAYU000 committed:
Update to_dagitty to handle o

ankurankan committed:
Merge branch 'dev' into feature/MAG_dagitty

ankurankan commented:
@JATAYU000 I think tests for `to_dagitty` method are missing because of which codecov is failing. Could you please add tests for it?

JATAYU000 committed:
Merge branch 'dev' into feature/MAG_dagitty

JATAYU000 committed:
Added some tests for more coverage

JATAYU000 committed:
Format tests

JATAYU000 committed:
Updated test assertion

JATAYU000 commented:
@ankurankan I have added tests for the coverage, should it be 100% because there few lines which are handling exceptions, are they supposed to be covered in the tests as well?

ankurankan committed:
Merge branch 'dev' into feature/MAG_dagitty

ankurankan reviewed (approved):
@JATAYU000  No, that's fine. Looks good to me now, thanks for the PR.
